pkg/kf/cli/prompts.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/manifoldco/promptui"
+)
+
+type interactiveModeType struct{}
+
+// SetInteractiveMode returns a context that marks it as interactive. This
+// allows prompts to make decisions based on if the user wants to interact
+// with the CLI interactively or not.
+func SetInteractiveMode(ctx context.Context, interactive bool) context.Context {
+	return context.WithValue(ctx, interactiveModeType{}, interactive)
+}
+
+// GetInteractiveMode returns true if SetInteractiveMode has been set to true
+// for the given context.
+func GetInteractiveMode(ctx context.Context) bool {
+	i, _ := ctx.Value(interactiveModeType{}).(bool)
+	return i
+}
+
+// interactiveModeDisabledErr is returned by a prompt if InteractiveMode is
+// not enabled but a prompt is still requested.
+type interactiveModeDisabledErr struct {
+	message string
+}
+
+// Error implements the error interface.
+func (e interactiveModeDisabledErr) Error() string {
+	return e.message
+}
+
+// IsInteractiveModeErr returns true if the given error is because a prompt
+// was requested when interactive mode was disabled.
+func IsInteractiveModeErr(err error) bool {
+	_, ok := err.(interactiveModeDisabledErr)
+	return ok
+}
+
+// SelectPrompt prompts the user to select from the given items. It uses
+// Searcher and properly colors the label. It fails if the context is not in
+// interactive mode.
+func SelectPrompt(
+	ctx context.Context,
+	label string,
+	items ...string,
+) (int, string, error) {
+	if !GetInteractiveMode(ctx) {
+		return 0, "", interactiveModeDisabledErr{message: label}
+	}
+
+	p := promptui.Select{
+		Label:             LabelColor.Sprint(label),
+		StartInSearchMode: true,
+		Searcher:          Searcher(items),
+		Items:             items,
+	}
+	return p.Run()
+}
+
+// SelectYesNo promts the user to select between yes and no. It will return
+// true if the user selects "yes", and false otherwise. It always returns yes
+// in non-interactive mode.
+func SelectYesNo(ctx context.Context, label string) (bool, error) {
+	if !GetInteractiveMode(ctx) {
+		return true, nil
```

Comment:
This doesn't feel very safe to me, for example this flag is used in delete today, I'd rather crash or hang forever rather than accidentally delete something the user needed; but in other cases the logic might be flipped.

Reply:
Fair point.

How would you feel about adding a default to it? That way the dev using the function could make a sane choice?

Reply:
SGTM